src/librustc/diagnostics.rs

```diff
@@ -335,10 +403,314 @@ This error indicates that an attempt was made to divide by zero (or take the
+can only be accessed behind a pointer. Thus, if we have a trait like the
+following:
+
+```
+trait Foo where Self: Sized {
+
+}
+```
+
+we cannot create an object of type `Box<Foo>` or `&Foo` since in this case
+`Self` would not be `Sized`.
+
+Generally `Self : Sized` is used to indicate that the trait should not be used
+as a trait object. If the trait comes from your own crate, consider removing
+this restriction.
+
+### Method references the `Self` type in its arguments or return type
+
+This happens when a trait has a method like the following:
+
+```
+trait Trait {
+    fn foo(&self) -> Self;
+}
+impl Trait for String {
+    fn foo(&self) -> Self {
+        "hi".to_owned()
+    }
+}
+
+impl Trait for u8 {
+    fn foo(&self) -> Self {
+        1
+    }
+}
+```
+
+In such a case, the compiler cannot predict the return type of `foo()` in a case
+like the following:
+
+```
+fn call_foo(x: Box<Trait>) {
+    let y = x.foo(); // What type is y?
+    // ...
+}
+```
+
+If the offending method isn't actually being called on the trait object, you can
+add a `where Self: Sized` bound on the method:
+
+```
+trait Trait {
+    fn foo(&self) -> Self where Self: Sized;
+    // more functions
+}
+```
+
+Now, `foo()` can no longer be called on the trait object, but you will be
+allowed to call other trait methods and construct the trait objects. With such a
+bound, one can still call `foo()` on types implementing that trait that aren't
+behind trait objects.
+
+### Method has generic type parameters
+
+As mentioned before, trait objects contain pointers to method tables. So, if we
+have
+
+```
+trait Trait {
+    fn foo(&self);
+}
+impl Trait for String {
+    fn foo(&self) {
+        // implementation 1
+    }
+}
+impl Trait for u8 {
+    fn foo(&self) {
+        // implementation 2
+    }
+}
+// ...
+```
+
+at compile time a table of all implementations of `Trait`, containing pointers
+to the implementation of `foo()` would be generated.
+
+This works fine, but when we the method gains generic parameters, we can have a
+problem.
+
+Usually, generic parameters get _monomorphized_. For example, if I have
+
+```
+fn foo<T>(x: T) {
+    // ...
+}
+```
+
+the machine code for `foo::<u8>()`, `foo::<bool>()`, `foo::<String>()`, or any
+other type substitution is different. Hence the compiler generates the
+implementation on-demand. If you call `foo()` with a `bool` parameter, the
+compiler will only generate code for `foo::<bool>()`. When we have additional
+type parameters, the number of monomorphized implementations the compiler
+generates does not grow drastically, since the compiler will only generate an
+implementation if the function is called with hard substitutions.
+
+However, with trait objects we have to make a table containing _every object
+that implements the trait_. Now, if it has type parameters, we need to add
+implementations for every type that implements the trait, bloating the table
+quickly.
```

Comment:
Codegen happens for monomorphic functions not marked with `#[inline]` regardless of use, and for everything else on-demand.
```rust
pub fn opaque_magic() -> Box<Trait> {
    struct MyImpl;
    impl Trait for MyImpl {...}
    box MyImpl
}
```
The example above will always be codegen'd in the crate where it is defined, and it would return `(0x1, vtable_ptr)`.